torch/_inductor/codegen/cpp_utils.py

```diff
@@ -330,77 +413,105 @@ def get_dtype(name):
 
         self.exit_stack.enter_context(patch.object(V.graph, "get_dtype", get_dtype))
 
-        original_input = self.kernel.args.input
+        original_input = self.kernel_args.input
 
         def input(name):
             if name in self.local_buffers:
                 return name
             return original_input(name)
 
-        self.exit_stack.enter_context(patch.object(self.kernel.args, "input", input))
+        self.exit_stack.enter_context(patch.object(self.kernel_args, "input", input))
 
-        original_output = self.kernel.args.output
+        original_output = self.kernel_args.output
 
         def output(name):
             if name in self.local_buffers:
                 return name
             return original_output(name)
 
-        self.exit_stack.enter_context(patch.object(self.kernel.args, "output", output))
+        self.exit_stack.enter_context(patch.object(self.kernel_args, "output", output))
+
+        # Set current LocalBufferContext into V
+        self.exit_stack.enter_context(V.set_local_buffer_context(self))
 
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         self.local_buffers.clear()
         self.exit_stack.__exit__(exc_type, exc_val, exc_tb)
 
-    def add_local_buffer(self, buffer: ir.Buffer):
-        assert buffer.get_name() not in self.local_buffers
-        self.local_buffers[buffer.get_name()] = buffer
+    def add_local_buffer(
+        self, local_buffer: ir.Buffer, global_buffer: Optional[ir.Buffer] = None
+    ):
+        assert local_buffer.get_name() not in self.local_buffers
+        self.local_buffers[local_buffer.get_name()] = local_buffer
+        if global_buffer:
+            self.local_to_global[local_buffer.get_name()] = global_buffer
+
+            def should_allocate():
+                assert isinstance(global_buffer, ir.Buffer)
+                # Never resue this buffer, sinice it's not real allocated.
+                V.graph.never_reuse_buffers.add(global_buffer.get_name())
+                # Mark this buffer as freed, so that we don't free it later.
+                V.graph.wrapper_code.freed.add(global_buffer.get_name())
+                # Avoid allocation of this Global Buffer
+                return False
+
+            assert isinstance(global_buffer, ir.Buffer)
+            # Patch the should_allocate of global buffer to avoid allocation
+            self.exit_stack.enter_context(
+                patch.object(global_buffer, "should_allocate", should_allocate)
```

Comment:
Can you not just add the global buffer to `removed_buffers`?

Reply:
Thanks for the comment. Change the implementation by adding `global_buffer` into `V.graph.removed_buffers`.